crypto/merkle/simple_map.go

```diff
@@ -66,23 +65,17 @@ func (sm *simpleMap) KVPairs() cmn.KVPairs {
 // then hashed.
 type KVPair cmn.KVPair
 
-func (kv KVPair) Hash() []byte {
-	hasher := tmhash.New()
-	err := encodeByteSlice(hasher, kv.Key)
-	if err != nil {
-		panic(err)
-	}
-	err = encodeByteSlice(hasher, kv.Value)
-	if err != nil {
-		panic(err)
-	}
-	return hasher.Sum(nil)
+func (kv KVPair) Bytes() []byte {
+	val := make([]byte, len(kv.Key)+len(kv.Value))
+	copy(val, kv.Key)
+	copy(val[len(kv.Key):], kv.Value)
```

Comment:
This means that, e.g. key = "Hello", value = "World" will hash to the same as key = "Hell", value = "oWorld". Isn't this concerning?

Reply:
linking a relevant discussion here: https://github.com/tendermint/tmlibs/pull/211#issuecomment-391102658

Reply:
OK, what I described above doesn't work, as the value get's hashed separately (in simpleMap):
https://github.com/tendermint/tendermint/blob/0bec20a1e3b716add4d105c8576b4dadd9339109/crypto/merkle/simple_map.go#L23-L36

If not one could simply do this and both hashes (of db and db2) would be the same:
```go
                db := newSimpleMap()
		db.Set("Hello", strHasher("World"))

		db2 := newSimpleMap()
		db2.Set("Hell", strHasher("oWorld"))
```

Reply:
Good point, we should change do what your describing!

Reply:
In #2595 I've opted to preserve the same map implementation there. I don't think we should hash the value, as I noted in #2593.

However we currently amino encode every leaf. So I think the optimal solution here from a simplcitiy POV would be to amino encode the KV pair struct, so no special casing is required on the proof side. I'll open a new issue for this.